### DISCUSSION TITLE
New Field Type: Address / Map

### DISCUSSION THREAD

**Discussion by sc0ttkclark** [pods-framework/pods#6628] [Ideas]
Supports Google Maps JS / Static maps and Geocoding; Support for Bing later on

PR & Task List #3634

# Old:
<details>

## Task List
- [x] Component
- [x] Custom Field Type
- [x] Admin Options for Component
- [x] Admin Options for Fields
- [x] Input styles for Address field group and Map
- [x] Output functions for address and maps
- [ ] UI method for when listed table item list (PodsUI)
- [x] Zip Code field that allows leading zeros (international)

## Additional tasks
- [x] Fix integration of the Google Maps API
- [x] Display field options (backend and frontend)
- [ ] Make sure code is set for easy integration of possible other map types (OpenStreetMap for example)
- [ ] Loads of testing with different configurations and switching between them
- [ ] Test with Pods templates and other functions
</details>

**Comment by pglewis** [comment]
http://www.youtube.com/watch?v=E_7j9kl4ZSI
http://www.advancedcustomfields.com/add-ons/google-maps/

**Comment by sc0ttkclark** [comment]
User searches for an address, that address and it's coordinates are saved in serialized array of data.

**Comment by pglewis** [comment]
Feature list: 
- Geolocation from address (done)
- lat/lng (done)
- Drag/drop marker (done)
- Custom marker icon
- Starting zoom level

Output format: Map Image, or Map JS

Other potential options: 
? map type (roadmap, terrain,...)
? sensor support
? marker title
? info window

**Comment by pglewis** [comment]
Phase II ideas: 
- Fusion Table layer integration
- KML layer integration
- directions

**Comment by pglewis** [comment]
Resource for generating shadows and shape coords for custom icons: 

http://powerhut.co.uk/googlemaps/custom_markers.php

**Comment by benfavre** [comment]
@pglewis not sure how you expect different output types to work and i think it's out of the scope of an input field.
This should probably only return lat long for 1 marker as does ACF

**Comment by pglewis** [comment]
Supporting both the Google image API and the JS API for the output map is Scott's idea.  The plan is, I believe, to have field() return an array with data (lat/lng, address, whatever else) and display() will automagically output a map and marker.  

An extra option to specify which API to use for the output when using display() doesn't seem too far out of scope to me, and I'm a fan of options over not.  It doesn't stop you from grabbing the lat/lng and doing the map yourself if you prefer.  

My question is how fine-grained the option should be?  Record level, field level, or global?

**Comment by pglewis** [comment]
I really want sensor support, too.  Displaying a map that can give me one-click directions from my phone?  Sexy.

**Comment by sc0ttkclark** [comment]
Right, we also want an output option to just output the full address

**Comment by sc0ttkclark** [comment]
This is coming along pretty well, admin options, field type, and component are all good. Need to work on Input and Display.

**Comment by sc0ttkclark** [comment]
Screenshot of field options:
![Screen Shot 2013-01-22 at 10 51 51 PM](https://f.cloud.github.com/assets/709662/89030/b3cea878-6518-11e2-8792-2509fd814a6b.png)

Screenshot of the new menu item to manage maps settings:
![Screen Shot 2013-01-22 at 10 52 38 PM](https://f.cloud.github.com/assets/709662/89033/c1497118-6518-11e2-94fa-eeebaeb0468b.png)

**Comment by sc0ttkclark** [comment]
More screenshots of latest updates:

![Component Settings](http://f.cl.ly/items/01270T323J1E3Q0s2m37/Screen%20Shot%202013-01-24%20at%2012.55.49%20PM.png)

![Field Options](http://f.cl.ly/items/2N0f0F1V2u100i061R13/Screen%20Shot%202013-01-24%20at%2012.56.48%20PM.png)

**Comment by imzeeshan** [comment]
Any update on this issue ?

**Comment by sc0ttkclark** [comment]
This is still on track to be one of our first external components, available on our new site. I'll make an effort to try and get it done as soon as I can, but there are a few other things ahead of it in terms of priority at the moment (namely #109 and #81)

**Comment by sc0ttkclark** [comment]
There is now a bounty on this issue, the admin UI is done, we just need someone to help us build the input field (for use in the form) and the output. Some of this may already be done based on @pglewis's work, but need someone to bring it home for us!

https://www.bountysource.com/issues/865545-new-field-type-address-map

**Comment by ghost** [comment]
I am keen to help out on this, where could I get the existing code? @pglewis

**Comment by jasonday** [comment]
Any movement on this? What specifically needs to be done, as I can assist.

**Comment by ghost** [comment]
I was hoping to have it done by now, but have been backed up by client work. It's the form inputs and HTML output that is remaining as far as I know. @sc0ttkclark can provide more details.

**Comment by sc0ttkclark** [comment]
@unknownnf is going to have a go at this, he'll update everyone on progress. We'll get anyone who's interested in testing, beta testing access if they have the time to test.

**Comment by ryanyam** [comment]
Looking forward to this release. It seems as if there's activity again on this issue.

**Comment by AlexandreNarten-zz** [comment]
This field type looks awesome. Loving the screenshots, gives a good idea of the functionality. Any news on this?

**Comment by Shelob9** [comment]
@AlexandreNarten At this point this field type is not a focus for us. We need to focus our limited resources on completing Pods 3.0. 

You should check out this post for an interesting way to work with Maps + Pods:
http://interi.org/2014/04/pods-maps/

**Comment by sc0ttkclark** [comment]
https://github.com/pods-framework/pods-address-maps

**Comment by sc0ttkclark** [comment]
Any update on this @unknownnf

**Comment by lpawlak** [comment]
How to install code from https://github.com/pods-framework/pods-address-maps ?

**Comment by Shelob9** [comment]
@Ipawlak That code is only compatible with Pods 3.0, which is not yet in a functional beta.

**Comment by JoryHogeveen** [comment]
@sc0ttkclark Should this field be a separate plugin or included within Pods?
I ask this because the rep you've created is for a separate plugin.
https://github.com/pods-framework/pods-address-maps

**Comment by JoryHogeveen** [comment]
Err, since when does commenting result in an assignment removal?

**Comment by JoryHogeveen** [comment]
Got the code from @sc0ttkclark compatible and working with the 2.7 branche so time to look at the features etc!

[task list moved to top of summary issue]

**Comment by quasel** [comment]
Status? Map Field is one of the most asked questions these days ;)

**Comment by JoryHogeveen** [comment]
Doing as much as I can. It's working but can have a lot of improvements. Plus, it's not integrated in DFV structure yet.
See PR for status

**Comment by quasel** [comment]
Don't worry - just came up in a bunch of conversations 👍 - any idea about a timeframe? I know Scott is a busy man! ;)

**Comment by sc0ttkclark** [comment]
Mostly testing I think?

**Comment by JoryHogeveen** [comment]
See tasklist, more than just testing ;)

**Comment by jimtrue** [comment]
@JoryHogeveen This is separate from the Address Field in #3634, correct?

**Comment by JoryHogeveen** [comment]
This is the attached issue. But the todolist is maintained in the PR.

**Comment by jimtrue** [comment]
Duh, you would think I'd notice that big Pull Request tag at the top huh? Think I got a little confused when I saw two task lists ;) Thanks for the clarification @joryhogeveen

**Comment by quasel** [comment]
- ACF Output:  The data saved and returned is an array containing the markers lat, lng, and address.
- Toolset: Stores Same data as ACF details - https://wp-types.com/documentation/customizing-sites-using-php/functions/
- MetaBox:  Stores Address separate and has a map field that takes the address_field ID as input https://docs.metabox.io/fields/map/#displaying-the-map - Array keys are latitude and longitude ... 
- CMB2 no address field

**Comment by qrahaman** [comment]
Hello,

Just following up on the Slack support discussion.  I asked whether or not this field would also auto-generate the location data; such as, city, country, and latitude and longitude.  I was told it would which is fantastic.  

In my particular case, my users are effectively playing a visual measurement game on their device, mobile or desktop, and upon completion of the game, a post/pod is submitted automatically.  The submission contains a whole host of data.  I would like their location to be part of that submission (without their having to either input that data or asked for approval to submit location data) since it will be used for filtering their scores by location and used to apply a marker to a global map.  

It would be awesome if it could be integrated with FacetWP.  Is there a projected delivery date for this field type addition?

**Comment by vesna-droid** [comment]
It would be cool if there would be a display option of a map with flags for every location entered for some post type or taxonomy or user. To use as a one map showing all the locations of some event or taxonomy of events or locations of users origins.

**Comment by JoryHogeveen** [comment]
@vesna-droid 
While it's not implemented yet and only available through code.
This is my current implementation for a grouped map:
https://github.com/JoryHogeveen/pods-address-maps/blob/master/pods-maps/pods-maps.php#L67-L88

You can pass multiple values into the first parameter (array format) and it will show all of these on one single map.

**Comment by vesna-droid** [comment]
Excellent. Altough I'm a beginner in php coding and could use an example of passing actual addresses (from meta custom fields of posts or users) as array or even hire someone willing to establish this at our web place... Feel free to send me examples if you have them and willing to share for purpose of my learning...

**Comment by 1HabD** [comment]
Hi, 

I use Pods with Elementor and Dynamic.ooo. In order to get the Dynamic-people to support this field, the told me that: 

_- The PODS integration only allows you to view the map on the backend side. The data that comes to us is just a string, while ACF sends the data of the map object._ 

So, do you see any opportunity to output the data as an "map object"?
--

**Reply by JoryHogeveen** [reply]
Hi @1HabD 
This is a feature that isn't done yet so the actual implementation is still to-do.
However, our display options will definitely support formatted outputs as with our other fields.
